r1cs/gadgets/std/src/fields/nonnative/params.rs

```diff
@@ -1,91 +1,169 @@
 use crate::fields::nonnative::NonNativeFieldParams;
 
 /// The surfeit used for finding the optimal parameters
-pub(crate) const SURFEIT:u32 = 10; 
+pub(crate) const SURFEIT:u32 = 10;
 
 /// Obtain the parameters from a `ConstraintSystem`'s cache or generate a new one
 #[must_use]
-pub const fn get_params(target_field_size: usize, base_field_size: usize) -> NonNativeFieldParams {
+pub const fn get_params(target_field_prime_length: usize, base_field_prime_length: usize, pseudo_mersenne_const: Option<u64>) -> NonNativeFieldParams {
     let (num_of_limbs, limb_size, _) =
-        find_parameters(base_field_size, target_field_size);
+        find_parameters(target_field_prime_length, base_field_prime_length, pseudo_mersenne_const);
     NonNativeFieldParams {
         num_limbs: num_of_limbs,
         bits_per_limb: limb_size,
     }
 }
 
+const fn ceil_log2(value: u128) -> usize{
+    let mut result = std::mem::size_of::<u128>() * 8 -
+        value.leading_zeros() as usize;
+    if value == 2u128.pow((result - 1) as u32) { result -= 1 }
+    result
+}
+
+// This function estimates the number of constrains for a mul+reduce when the target prime field is
+// pseudo-mersenne and field elements are represented with num_limbs base field elements of size
+// bits_per_limb. The function returns the estimated number of constraints if the security bound for
+// the given parameters holds, None otherwise
+const fn estimate_cost_for_pseudomersenne(target_field_prime_length: usize, base_field_prime_length: usize, pseudo_mersenne_const: u64, bits_per_limb: usize, num_limbs: usize) -> Option<usize> {
+    let num_adds_plus_one = 2u128.pow(SURFEIT);
+    let capacity = base_field_prime_length  - 1;
+    let h = bits_per_limb*num_limbs - target_field_prime_length;
+    // cost of mul
+    let mut constraints = num_limbs*num_limbs;
+    // compute surfeit_prod in a compatible way with const function:
+    // surfeit_prod = 1 + log(num_limbs*(num_adds+1)*(num_adds+1)*(c*2^h+1)*2^bits_per_limb + 2)
+    // surfeit_prod may overflow u128 if bits_per_limb is high
```

Comment:
In the function get_params you write:
surfeit(product) = log(num_limbs * (num_add(L)+1) * (num_add(R) + 1) * (c+1) * 2^bits_per_limb),
i.e.
product < num_limbs * (num_add(L)+1) * (num_add(R) + 1) * (c+1) * 2^bits_per_limb
for the general case (it should be **c*h + 1**), 
prod_limb < num_limbs * (num_add(L)+1) * (num_add(R) + 1) * (c+1) * 2^(2*bits_per_limb)
This implies that
prod_limb_mersenne <= (1 + c*h) * prod_limb.
So, we should have that
surfeit_prod = log(num_limbs*(num_adds+1)*(num_adds+1)*(c*2^h+1)*2^bits_per_limb).
For the issue of overflowing 128 bits you can take out of the logarithm 2^bits_per_limb.
So, surfeit_prod = bits_per_limb + log(num_limbs*(num_adds+1)*(num_adds+1)*(c*2^h+1))

Reply:
Uhm, this is what the code currently does (besides the fact that you also need to handle the possible overflow of 2^h), so I am not sure what change you are suggesting :)

Reply:
I do not understand the +1 outside the log and the +2 inside the log.
You are right that theorically there can be overflows with u128, but i guess that in practice it does not happen if you take out 2^bits_per_limb and you choose a tiny value of h.

Reply:
The +1 and the +2 come from the security bounds on the reduction. The naming of the variable is maybe misleading: that surfeit is not the one of the product, but the one needed to verify the security bound of a multiplication, which must produce a reducible result.
In general h can be as big as bits_per_limb, and since the function estimates the cost for bits_per_limb up to capacity, then also 2^h could overflow. I wrote the code to deal with all the cases, it can be simplified if we put an upper bound to h as you pointed out. So the question is: shall we place a reasonable upper bound on h? How do we establish such an upper bound?